api/apiroutes.go

```diff
@@ -1099,3 +1101,38 @@ func (c *appContext) getBlockHashCtx(r *http.Request) string {
 	}
 	return hash
 }
+
+// ZeroAddrDenier https://github.com/decred/dcrdata/issues/358
+func (c *appContext) ZeroAddrDenier(next http.Handler) http.Handler {
```

Comment:
This can stay in `package middleware`, no need to duplicate code.  More importantly, no need for an RPC call to figure out what net you're running on.  Just pass the params as I suggested in the last review.

So instead of this being a bare middleware, you'll have a factory func:

```go
func (c *appContext) ZeroAddrDenier(params *chaincfg.Params) func(http.Handler) http.Handler {
    deniedAddr := apitypes.NewZeroAddressDenial(params)
    return func(next http.Handler) http.Handler {
        ... // do stuff with deniedAddress
        next.ServeHTTP(w, r.WithContext(r.Context()))
    }
}
```

Then my comment about `app.ZeroAddrDenier(params)` should make sense.

Reply:
appContext this is internal  struct of api package. I can not use it in insight package as well as to use it in middleware as you suggested on screen.

Moreover, api imports middleware package. When I try to import api package in middleware, I got cycle import error.

Reply:
Yeah, I forgot to remove the `(c *appContext)` receiver in my example.  The point is you don't need `c`.  It'll be a regular function in the middleware package.